code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm

```diff
@@ -0,0 +1,562 @@
+	ghost.alpha = initial(ghost.alpha) //Become opaque
+	ghost.remove_filter("wraith_phase_shift") //Cool filter begone
+
+	playsound(owner, "sound/effects/phasein.ogg", 25, 0, 1)
+
+	ghost.add_filter("wraith_phase_shift_windup_1", 3, list("type" = "wave", 0, 0, size=rand()*2.5+0.5, offset=rand())) //Cool filter appear
+	ghost.add_filter("wraith_phase_shift_windup_2", 3, list("type" = "wave", 0, 0, size=rand()*2.5+0.5, offset=rand())) //Cool filter appear
+	animate(ghost.get_filter("wraith_phase_shift_windup_1"), x = 60*rand() - 30, y = 60*rand() - 30, size=rand()*2.5+0.5, offset=rand(), time = 0.25 SECONDS, loop = -1, flags=ANIMATION_PARALLEL)
+	animate(ghost.get_filter("wraith_phase_shift_windup_2"), x = 60*rand() - 30, y = 60*rand() - 30, size=rand()*2.5+0.5, offset=rand(), time = 0.25 SECONDS, loop = -1, flags=ANIMATION_PARALLEL)
+	addtimer(CALLBACK(ghost, /atom.proc/remove_filter, "wraith_phase_shift_windup_1"), 0.5 SECONDS)
+	addtimer(CALLBACK(ghost, /atom.proc/remove_filter, "wraith_phase_shift_windup_2"), 0.5 SECONDS)
+
+	var/current_turf = get_turf(ghost)
+
+	if(isclosedturf(current_turf) || isspaceturf(current_turf)) //So we rematerialized in a solid wall/space for some reason; Darwin award winner folks.
+		to_chat(ghost, "<span class='highdanger'>As we idiotically rematerialize in an obviously unsafe position, we revert to where we slipped out of reality at great cost.</span>")
+		ghost.adjustFireLoss((ghost.health * 0.5), TRUE) //Lose half of our health
+		ghost.Paralyze(5 SECONDS, xeno_adjust_stun = FALSE) //That oughta teach them.
+		ghost.forceMove(starting_turf)
+		teleport_debuff_aoe(ghost) //Debuff when we reappear
+		starting_turf = null
+		return
+
+	var/distance = get_dist(current_turf, starting_turf)
+	var/phase_shift_stun_time = clamp(distance * 0.1 SECONDS, 0.5 SECONDS, 3 SECONDS) //Recovery time
+	ghost.ParalyzeNoChain(phase_shift_stun_time, xeno_adjust_stun = FALSE)
+	ghost.visible_message("<span class='warning'>[ghost] form wavers and becomes opaque.</span>", \
+	"<span class='highdanger'>We phase back into reality, our mind reeling from the experience. We estimate we will take [phase_shift_stun_time * 0.1] seconds to recover!</span>")
+
+	starting_turf = null
+
+/datum/action/xeno_action/phase_shift/on_cooldown_finish()
+
+	to_chat(owner, "<span class='xenonotice'>We are able to fade from reality again.</span>")
+	owner.playsound_local(owner, 'sound/effects/xeno_newlarva.ogg', 25, 0, 1)
+	return ..()
+
+
+// ***************************************
+// *********** Resync
+// ***************************************
+/datum/action/xeno_action/resync
+	name = "Resync"
+	action_icon_state = "resync"
+	mechanics_text = "Resynchronize with realspace, ending Phase Shift's effect."
+	cooldown_timer = 1 SECONDS //Token for anti-spam
+	keybind_signal = COMSIG_XENOABILITY_RESYNC
+
+
+/datum/action/xeno_action/resync/action_activate()
+	. = ..()
+	if(!owner.get_filter("wraith_phase_shift"))
```

Comment:
why not check the flags instead

Reply:
it's less precise mainly

Reply:
Use the flags (such as incorporeal) they are much cheaper to check